### PR TITLE
clean up lib functions

### PR DESCRIPTION
**Pull request by Maxez**
## Description

- every function in /lib starts with `lib_` prefix,
- one main /lib header (`lib.h`) includes all other headers,
- varargs moved from hal/types.h to lib/stdarg.h,
- added sprintf() function,
- strtoul() and strtol() return unsigned long and long,

## Motivation and Context

Code clean up.
[JIRA: PD-156](https://jira.phoenix-rtos.com/browse/PD-156)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: lib/log.h [hunk 1]
```diff
@@ -33,4 +25,8 @@ extern int log_getEcho(void);
 extern void log_setEcho(int val);
 
 
+#define log_info(fmt, ...)  do { if (log_getEcho()) lib_printf(fmt, ##__VA_ARGS__); } while (0)
+#define log_error(fmt, ...) do { lib_printf(CONSOLE_RED fmt CONSOLE_NORMAL, ##__VA_ARGS__); } while (0)
```
[the remark this answers is not in the document]
Comment (Maxez) [reply lib/log.h]: Fixed.

File: lib/strtoul.c [hunk 2]
```diff
@@ -50,13 +51,13 @@ unsigned int lib_strtoul(char *nptr, char **endptr, int base)
...
 
-int lib_strtol(char *nptr, char **endptr, int base)
+long lib_strtol(char *nptr, char **endptr, int base)
 {
 	int sign = 1;
 
 	if (*nptr == '-') {
 		sign = -1;
-		++nptr;
+		nptr++;
```
Comment (agkaminski) [inline lib/strtoul.c:60]: Why this change? We preffer `++nptr`

File: lib/format.c [hunk 3]
```diff
@@ -452,9 +438,8 @@ static void format_sprintf_num(void *ctx, feedfunc feed, u64 num64, u32 flags, i
 	}
 	else if (flags & FLAG_OCT) {
 		if (flags & FLAG_64BIT) {
-			int i;
 			// 30 bits
-			for (i = 0; i < 10; ++i) {
+			for (i = 0; i < 10; i++) {
```
Comment (agkaminski) [inline lib/format.c:442]: We prefer ++i

File: lib/format.c [hunk 4]
```diff
@@ -411,36 +400,33 @@ static void format_sprintf_num(void *ctx, feedfunc feed, u64 num64, u32 flags, i
...
 
-	if (num64 == 0) {
+	if (!num64) {
 		*tmp++ = '0';
 	}
 	else if (flags & FLAG_HEX) {
 		if (flags & FLAG_64BIT) {
-			int i;
-			for (i = 0; i < 8; ++i) {
+			for (i = 0; i < 8; i++) {
```
Comment (agkaminski) [inline lib/format.c:419]: We prefer ++i

**Review by agkaminski — APPROVED** [review]

**Review by HBuczynski — CHANGES_REQUESTED** [review]
Please update authors in the files where you have made changes.

File: lib/strtoul.c (hunk 2 above)
Comment (Maxez) [reply lib/strtoul.c:60]: I think that in most of our codebase we use postincrementation (just my observation) so I tried to keep the code more consistent. Should we prefer one over the other when it doesn't change the logic?

File: lib/format.c (hunk 3 above)
Comment (Maxez) [reply lib/format.c:442]: And here is the answer for my question above :P I'll change it back to postincrement.

File: lib/format.c [hunk 5]
```diff
@@ -37,7 +36,7 @@
 
 #define GET_UNSIGNED(number, flags, args) \
 	do { \
-		if ((flags)&FLAG_64BIT) \
+		if ((flags) & FLAG_64BIT) \
```
[the remark this answers is not in the document]
Comment (Maxez) [reply lib/format.c:39]: We do want to use spaces around bit and operator.

File: lib/ctype.c [hunk 6]
```diff
@@ -1,9 +1,9 @@
 /*
  * Phoenix-RTOS
  *
- * phoenix-rtos loader
+ * Operating system loader
  *
- * ctype - code derived from libphoenix
+ * Character type
```
Comment (HBuczynski) [inline lib/ctype.c:6]: The comment that code is derived from libphoenix should be left to know its origin.
I think it will be important if the common part is shared between plo, kernel and libphoenix.

File: lib/ctype.h [hunk 7]
```diff
@@ -1,9 +1,9 @@
 /*
  * Phoenix-RTOS
  *
- * phoenix-rtos loader
+ * Operating system loader
  *
- * ctype - code derived from libphoenix
+ * Character type
```
Comment (HBuczynski) [inline lib/ctype.h:6]: The comment that code is derived from libphoenix should be left to know its origin.
I think it will be important if the common part is shared between plo, kernel and libphoenix.

File: lib/lib.h [hunk 8]
```diff
@@ -13,25 +13,41 @@
...
+#include "errno.h"
+#include "list.h"
+#include "log.h"
+#include "stdarg.h"
 
 
-#define min(a, b) ((a > b) ? b : a)
+#define min(a, b) ({ \
+	__typeof__ (a) _a = (a); \
+	__typeof__ (b) _b = (b); \
```
[the remark this answers is not in the document]
Comment (Maxez) [reply lib/lib.h]: Fixed.

File: lib/lib.h [hunk 9]
```diff
@@ -13,25 +13,41 @@
...
+	_a > _b ? _b : _a; \
+})
+
+
+#define max(a, b) ({ \
+	__typeof__ (a) _a = (a); \
+	__typeof__ (b) _b = (b); \
```
[the remark this answers is not in the document]
Comment (Maxez) [reply lib/lib.h]: Fixed.

File: lib/strtoul.c (hunk 2 above)
Comment (Maxez) [reply lib/strtoul.c]: Reverted back to preinrcementation.

File: lib/format.c (hunk 4 above)
Comment (Maxez) [reply lib/format.c]: Reverted back to ++i.

File: lib/format.c (hunk 3 above)
Comment (Maxez) [reply lib/format.c]: Fixed.

File: lib/ctype.c (hunk 6 above)
Comment (Maxez) [reply lib/ctype.c]: Restored libphoenix origin info.

File: lib/ctype.h (hunk 7 above)
Comment (Maxez) [reply lib/ctype.h]: Restored libphoenix origin info. Same for the rest of files that were marked so.

**Comment by Maxez** [comment]
> 
> 
> Please update authors in the files where you have made changes.

I've added myself to the authors list in files in which I modified the code (skipped files with fixed typos/comments).

**Review by HBuczynski — APPROVED** [review]

**Review by agkaminski — APPROVED** [review]
